index.d.ts

```diff
@@ -0,0 +1,438 @@
+export const allSettled: typeof RSVP.allSettled;
+export const hash: typeof RSVP.hash;
+export const race: typeof Promise.race;
+
+export function timeout(ms: number): Promise<void>;
+
+/**
+ * Use `waitForEvent` to pause the task until an event is fired.
+ * The event can either be a jQuery event or an `Ember.Evented` event, or any
+ * event system where the object supports `.on()`, `.one()` and `.off()`.
+ */
+export function waitForEvent(
+  object: EmberObject | EventTarget,
+  eventName: string
+): Promise<Event>;
+
+/**
+ * Use `waitForProperty` to pause the task until a property on an object
+ * changes to some expected value.
+ *
+ * This can be used for a variety of use cases, including synchronizing with
+ * another task by waiting for it to become idle, or change state in some
+ * other way.
+ *
+ * If you omit the callback, `waitForProperty` will resume execution when the
+ * observed property becomes truthy. If you provide a callback, it'll be
+ * called immediately with the observed property's current value, and multiple
+ * times thereafter whenever the property changes, until you return a truthy
+ * value from the callback, or the current task is canceled.
+ *
+ * You can also pass in a non-`function` value in place of the callback, in
+ * which case the task will continue executing when the property's value
+ * becomes the value that you passed in.
+ */
+export function waitForProperty<T extends object, K extends keyof T>(
+  object: T,
+  key: K,
+  callbackOrValue?: (value: T[K]) => boolean | any
+): Promise<void>;
+
+/**
+ * Use `waitForQueue` to pause the task until a certain run loop queue is
+ * reached.
+ */
+export function waitForQueue(queueName: string): Promise<void>;
+
+export function task<Args extends any[], R>(
+  taskFn: GeneratorFn<unknown, Args, R>
+): Task<Args, Exclude<R, Promise<any>>>;
+export function task<Args extends any[], R>(encapsulatedTask: {
+  perform: GeneratorFn<unknown, Args, R>;
+}): Task<Args, Exclude<R, Promise<any>>>;
+
+export function taskGroup(): TaskGroupProperty;
+
+interface CommonTaskProperty {
+  restartable: () => TaskProperty;
+  drop: () => TaskProperty;
+  keepLatest: () => TaskProperty;
+  enqueue: () => TaskProperty;
+  maxConcurrency: (n: number) => TaskProperty;
+  cancelOn: (eventName: string) => TaskProperty;
+  group: (groupName: string) => TaskProperty;
+}
+
+export interface TaskProperty extends CommonTaskProperty {
+  evented: () => TaskProperty;
+  debug: () => TaskProperty;
+  on: (eventName: string) => TaskProperty;
+}
+
+// eslint-disable-next-line @typescript-eslint/no-empty-interface
+export interface TaskGroupProperty extends CommonTaskProperty {}
+
+// Based on https://github.com/CenterForOpenScience/ember-osf-web/blob/7933316efae805e00723789809bdeb58a96a286a/types/ember-concurrency/index.d.ts
+
+/**
+ * Describes the state that the task instance is in.
+ * Can be used for debugging, or potentially driving some UI state.
+ */
+export enum TaskInstanceState {
```

Comment:
I'd avoid creating the `TaskInstance` and `TaskState` enums, since they suggest that _values_ (not just types) can be imported from [`ember-concurrency`](https://github.com/machty/ember-concurrency/blob/master/addon/index.js). If it's _necessary_ to make some utility types available to consumers, that should probably be done through a ghost module to avoid clashing with anything that might be exported as a value from the underlying library in the future

Reply:
Blame @chriskrycho 
https://gist.github.com/chriskrycho/9ed39076ccb698023e57e59080fedecd#file-ember-concurrency-d-ts-L4-L10
😜

Reply:
haha. I suspect the quality bar may be different different between Chris' personal gist and merging something in as official typings.

Reply:
That’s just crazy talk. :laughing: I agree here. Much as the enums are nice, if they can’t actually be imported we shouldn’t use them. Instead, we should just use literal types.